src/core/or/circuitpadding_machines.c

```diff
@@ -0,0 +1,351 @@
+  machine->states[CIRCPAD_STATE_BURST].
+      next_state[CIRCPAD_EVENT_PADDING_SENT] = CIRCPAD_STATE_BURST;
+  machine->states[CIRCPAD_STATE_BURST].
+      next_state[CIRCPAD_EVENT_PADDING_RECV] = CIRCPAD_STATE_BURST;
+}
+
+/* Setup the BURST state of the machine that hides client-side intro
+ * circuits. */
+static void
+setup_burst_state_for_hiding_intro_circuits(circpad_state_t *burst_state,
+                                            bool is_client)
+{
+  /* Token removal strategy for BURST state */
+  burst_state->token_removal = CIRCPAD_TOKEN_REMOVAL_CLOSEST;
+
+  /* Histogram for BURST state:
+   *
+   * We want a machine that will send up to 20 fake cells over a reasonable
+   * time frame, to obfuscate the fact that introduction circuits always send 4
+   * cells both ways */
+
+  /* Figure out the length of the BURST state so that it's randomized. See
+     definition of these constants for rationale. */
+  burst_state->length_dist.type = CIRCPAD_DIST_UNIFORM;
+  burst_state->length_dist.param1 = is_client ?
+    HS_MACHINE_PADDING_MINIMUM_CLIENT : HS_MACHINE_PADDING_MINIMUM_SERVICE;
+  burst_state->length_dist.param2 = is_client ?
+    HS_MACHINE_PADDING_MAXIMUM_CLIENT : HS_MACHINE_PADDING_MAXIMUM_SERVICE;
+
+  /* We mainly care about destroying circuit construction fingerprints. Here is
+   * an example intro circuit construction, where the timestamp granularity is
+   * 1 millisecond:
+   *
+   * IP circuit relay cells:
+   *  0: 11:56:59.703 -> EXTEND2
+   *  1: 11:57:00.074 <- EXTENDED2
+   *  2: 11:57:00.075 -> EXTEND2
+   *  3: 11:57:00.243 <- EXTENDED2
+   *  4: 11:57:17.406 -> EXTEND
+   *  5: 11:57:17.708 <- EXTENDED
+   *  6: 11:57:17.709 -> INTRODUCE1
+   *  7: 11:57:17.904 <- INTRODUCE_ACK
+   *
+   * So with about 200 milliseconds latency between each cell, we appropriately
+   * set the histogram edges to from 30 milliseconds to 100 milliseconds, to
+   * account for faster connections.
+   */
+
+  /* Histogram is: (30 msecs, 200 msecs, infinity). */
+  burst_state->histogram_len = 2;
+  burst_state->histogram_edges[0] = 30000;
+  burst_state->histogram_edges[1] = 200000;
+
+  /* Maximum number of tokens for the BURST state */
+  burst_state->histogram[0] = 1000;
+  burst_state->histogram_total_tokens = 1000;
+}
+
+/* Setup the BURST state of the machine that hides client-side rend
+ * circuits. If <b>is_client</b> is true, then this is the client-side part of
+ * this machine, otherwise it's the service-side of this machine. */
+static void
+setup_burst_state_for_hiding_rend_circuits(circpad_state_t *burst_state,
+                                           bool is_client)
+{
+  /* Token removal strategy for BURST state */
+  burst_state->token_removal =
+      CIRCPAD_TOKEN_REMOVAL_CLOSEST;
+
+  /* Histogram for BURST state:
+   *
+   * We want a machine that will send a bunch of fake cells over a reasonable
+   * time frame, to obfuscate the fact that introduction circuits always send 4
+   * cells both ways */
```

Comment:
This comment seems to be cut+paste from the introduction machine? Needs update.

Reply:
Fixed in 7dc7c02f13.